### PR TITLE
Feature/navigate attachments in gallery

### PR DESCRIPTION
**Pull request by PenghaiZhang**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [x] tests are included
- [x] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This PR adds the support for viewing previous/next attachments with Lightbox in Gallery mode. The function used to build the handler for `onNext` and `onPrevious` was focused on `Attachment` so it did not support `GalleryEntry` very well. So I refactored this function to support both.

During implementation, Sam found a bug where if you are on the first/last image and you press the left/right arrow button you will see a blank Lightbox. The fix is also added in this PR.

https://user-images.githubusercontent.com/47203811/121481993-052c2a00-ca10-11eb-82f1-182278a90282.mp4

### CODE REVIEW

File: react-front-end/tsrc/components/Lightbox.tsx [hunk 1]
```diff
@@ -140,22 +134,17 @@ const Lightbox = ({ open, onClose, config }: LightboxProps) => {
...
   const { src, title, mimeType, onPrevious, onNext } = lightBoxConfig;
 
-  const handleOnPrevious = useCallback(() => {
-    setContent(undefined);
-    onPrevious && setLightBoxConfig(onPrevious());
-  }, [onPrevious]);
-
-  const handleOnNext = useCallback(() => {
+  const handleNav = (getLightboxConfig: () => LightboxConfig) => {
     setContent(undefined);
```
Comment (PenghaiZhang) [inline react-front-end/tsrc/components/Lightbox.tsx:138]: Here is the cause for the bug. We set `content` to undefined before we get the new lightbox config. The fix is that we do nothing when the handlers (onNext / onPrevious) provided are undefined.

File: react-front-end/tsrc/modules/ViewerModule.ts [hunk 2]
```diff
@@ -187,39 +201,50 @@ export const getViewerDefinitionForAttachment = (
 };
 
 /**
- * Build a function to handler navigation between Lightbox attachments.
- * @param lightboxAttachments All attachments that can be viewed in Lightbox and their Viewer definitions.
- * @param attachmentIndex Index of the attachment to be viewed
+ * Build a function to handler navigation between Lightbox entries.
```
Comment (edalex-ian) [inline react-front-end/tsrc/modules/ViewerModule.ts:204]:
```suggestion
 * Build a function to handle navigation between Lightbox entries.
```

File: react-front-end/tsrc/search/components/GallerySearchResult.tsx [hunk 3]
```diff
@@ -110,16 +126,31 @@ const GallerySearchResult = ({ items }: GallerySearchResultProps) => {
...
+                true
+              ),
+              onPrevious: buildLightboxNavigationHandler(
+                lightboxEntries,
+                initialLightboxEntryIndex - 1,
+                true
+              ),
             },
           });
+        };
```
Comment (edalex-ian) [inline react-front-end/tsrc/search/components/GallerySearchResult.tsx:153]: I think this change - maybe the one before - has pushed `mapItemsToTiles()` over the edge of readability. I think perhaps some of the nested functions should be pushed out - such as `buildOnClickHandler()` - and we probably also probably want to minimise use of external variables to help - such as using `lightboxEntries`.

Refactor needed.

**Review by edalex-ian — CHANGES_REQUESTED** [review]
Looks good @PenghaiZhang , but I think we need one refactor before proceeding.

**Review by edalex-ian — APPROVED** [review]
Good stuff @PenghaiZhang . Merge on :+1:

**Pull request merged**